presto-main/src/main/java/com/facebook/presto/operator/aggregation/MapUnionSumResult.java

```diff
@@ -0,0 +1,321 @@
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.Type;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.facebook.presto.common.type.TypeUtils.isExactNumericType;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public abstract class MapUnionSumResult
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapUnionSumResult.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
+    private static final int EXPECTED_ENTRY_SIZE = 16;
+
+    protected final Type keyType;
+    protected final Type valueType;
+    protected final Adder adder;
+
+    public MapUnionSumResult(Type keyType, Type valueType, Adder adder)
+    {
+        this.keyType = requireNonNull(keyType, "keyType is null");
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        this.adder = adder;
+    }
+
+    abstract int size();
+    abstract void addKey(int i, BlockBuilder out);
+    abstract void appendValue(int i, BlockBuilder blockBuilder);
+    abstract boolean isValueNull(int i);
+    public abstract long getRetainedSizeInBytes();
+    abstract void addKeyToSet(TypedSet keySet, int i);
+    abstract int getPosition(TypedSet otherKeySet, int keyPosition);
+    abstract TypedSet buildKeySet();
+    abstract Block getValueBlock();
+    abstract int getValueBlockIndex(int i);
+
+    public static MapUnionSumResult create(Type keyType, Type valueType, Adder adder, Block mapBlock)
+    {
+        return new SingleMapBlock(keyType, valueType, adder, mapBlock);
+    }
+
+    public Type getKeyType()
+    {
+        return keyType;
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        BlockBuilder mapBlockBuilder = out.beginBlockEntry();
+        for (int i = 0; i < size(); i++) {
+            addKey(i, mapBlockBuilder);
+            appendValue(i, mapBlockBuilder);
+        }
+        out.closeEntry();
+    }
+
+    static void appendValue(Type valueType, Block block, int position, BlockBuilder blockBuilder)
+    {
+        if (block.isNull(position)) {
+            if (isExactNumericType(valueType)) {
+                valueType.writeLong(blockBuilder, 0L);
+            }
+            else {
+                valueType.writeDouble(blockBuilder, 0);
+            }
+        }
+        else {
+            valueType.appendTo(block, position, blockBuilder);
+        }
+    }
+
+    /**
+     * Add the values for corresponding keys from other to this.
+     */
+    public MapUnionSumResult unionSum(MapUnionSumResult other)
+    {
+        if (other instanceof KeySetAndValues &&
+                (this instanceof SingleMapBlock || size() < other.size())) {
+            // We can avoid create a new set by merging into the other one.
```

Comment:
why would merging into the other one avoid creating a new set?

Reply:
The typed set continues to grow and we don't need to create a new one.

Reply:
"avoid creating a new key set"?